### PR TITLE
Feature/orca 334

### PR DESCRIPTION
**Pull request by clappn**
## Summary of Changes

Addresses [ORCA-334](https://bugs.earthdata.nasa.gov/browse/ORCA-334)

## Changes

* Created new aws_iam_policy_document, aws_iam_role, and aws_iam_role_policy objects used for lambda execution
* Created an aws_iam_role_policy_attachment used for lambda deployment and attached it to the created aws_iam_role
* Set the role to be used by the extract_filepaths_for_granule lambda

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

- Test the lambda using the lambda execution UI
- Test the lambda using the OrcaRecoveryWorkflow stepfunction

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the necessary documentation (remove those that do not apply)
    - [x] CHANGELOG.md
- [x] My changes generate no new warnings
- [x] My code has passed security scanning
    - [x] Snyk
    - [x] git-secrets

### CODE REVIEW

File: modules/lambdas/main.tf [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # Local Variables
 locals {
+  tags         = merge(var.tags, { Deployment = var.prefix })
```
Comment (rizbih) [inline modules/lambdas/main.tf:3]: I think you will need to merge develop branch into this branch to pull in the latest changes. 

In the develop branch, it no longer has the local tags value.

File: modules/lambdas/main.tf [hunk 2]
```diff
@@ -146,6 +147,55 @@ resource "aws_lambda_function" "extract_filepaths_for_granule" {
 }
 
 
+data "aws_iam_policy_document" "assume_lambda_role_extract" {
+  statement {
+    principals {
+      type = "AWS"
+      identifiers = ["*"]
```
Comment (rizbih) [inline modules/lambdas/main.tf:154]: The identifier should be `["lambda.amazonaws.com"]`

See similar example here https://github.com/nasa/cumulus-orca/blob/4d6cff91690984c4a8b947d89a2115d1e0009301/modules/iam/main.tf#L17

File: modules/lambdas/main.tf [hunk 3]
```diff
@@ -146,6 +147,55 @@ resource "aws_lambda_function" "extract_filepaths_for_granule" {
 }
 
 
+data "aws_iam_policy_document" "assume_lambda_role_extract" {
+  statement {
+    principals {
+      type = "AWS"
```
Comment (rizbih) [inline modules/lambdas/main.tf:153]: Should the type be `AWS `or `Service`?
In similar example in orca, it uses Service
https://github.com/nasa/cumulus-orca/blob/4d6cff91690984c4a8b947d89a2115d1e0009301/modules/iam/main.tf#L16

File: modules/lambdas/main.tf [hunk 4]
```diff
@@ -146,6 +147,55 @@ resource "aws_lambda_function" "extract_filepaths_for_granule" {
...
+
+  }
+}
+
+# Permissions needed to run the function successfully
+data "aws_iam_policy_document" "extract_filepaths_for_granule_policy_document" {
+  statement {
+    actions = [
+      "lambda:Invoke*",
+      "s3:PutObject*"
```
Comment (rizbih) [inline modules/lambdas/main.tf:166]: are we adding anything to a bucket using that lambda? If not, I think this can be removed.

**Review by rizbih — CHANGES_REQUESTED** [review]
A few things to look into.

File: modules/lambdas/main.tf (hunk 1 above)
Comment (clappn) [reply modules/lambdas/main.tf]: I changed this back. The original intention was to keep tags self-contained so we wouldn't need to rely on a .tfvars file but it shouldn't matter either way

File: modules/lambdas/main.tf [hunk 5]
```diff
@@ -146,6 +147,55 @@ resource "aws_lambda_function" "extract_filepaths_for_granule" {
...
+
+resource "aws_iam_role_policy" "extract_filepaths_for_granule_policy" {
+  name   = "${var.prefix}_extract_filepaths_for_granule_policy"
+  role   = aws_iam_role.extract_filepaths_for_granule_iam_role.id
+  policy = data.aws_iam_policy_document.extract_filepaths_for_granule_policy_document.json
+}
+
+
+# Attaches permissions needed to deploy the Lambda
+resource "aws_iam_role_policy_attachment" "extract_filepaths_for_granule_iam_policy_attachment" {
```
[the remark this answers is not in the document]
Comment (clappn) [reply modules/lambdas/main.tf]: Removed this and moved the actions that the attachment would give to the `extract_filepaths_for_granule_policy_document` policy document

File: modules/lambdas/main.tf (hunk 2 above)
Comment (clappn) [reply modules/lambdas/main.tf]: Fixed this, the AWS tech said I needed to have both "lambda.amazonaws.com" and "states.amazonaws.com" set here

File: modules/lambdas/main.tf (hunk 4 above)
Comment (clappn) [reply modules/lambdas/main.tf]: The lambda does not work at all if "s3:PutObject" isn't defined, I tried using "s3:GetObject" but the function and step function immediately fail.

File: modules/lambdas/main.tf (hunk 1 above)
Comment (rizbih) [reply modules/lambdas/main.tf]: looks good now

File: modules/lambdas/main.tf (hunk 2 above)
Comment (rizbih) [reply modules/lambdas/main.tf]: sounds good

File: modules/lambdas/main.tf (hunk 4 above)
Comment (rizbih) [reply modules/lambdas/main.tf]: okay then that makes sense to add it

**Review by rizbih — APPROVED** [review]
looks good to me now. Hope you did some testing by manually triggering the lambda from AWS console to see if it gives any weird permission error.

**Comment by clappn** [comment]
Yep, tested using manual triggers of the lambda and step function, and several deployments/destroys to ensure the lambda and dependencies would be created and run without error

**Pull request merged**